plumbing/format/commitgraph/file.go

```diff
@@ -1,259 +1,350 @@
-	} else if parent1 != parentNone {
-		parentIndexes = []int{int(parent1 & parentOctopusMask)}
-	}
-
-	parentHashes, err := fi.getHashesFromIndexes(parentIndexes)
-	if err != nil {
-		return nil, err
-	}
-
-	return &CommitData{
-		TreeHash:      treeHash,
-		ParentIndexes: parentIndexes,
-		ParentHashes:  parentHashes,
-		Generation:    int(genAndTime >> 34),
-		When:          time.Unix(int64(genAndTime&0x3FFFFFFFF), 0),
-	}, nil
-}
-
-func (fi *fileIndex) getHashesFromIndexes(indexes []int) ([]plumbing.Hash, error) {
-	hashes := make([]plumbing.Hash, len(indexes))
-
-	for i, idx := range indexes {
-		if idx >= fi.fanout[0xff] {
-			return nil, ErrMalformedCommitGraphFile
-		}
-
-		offset := fi.oidLookupOffset + int64(idx)*20
-		if _, err := fi.reader.ReadAt(hashes[i][:], offset); err != nil {
-			return nil, err
-		}
-	}
-
-	return hashes, nil
-}
-
-// Hashes returns all the hashes that are available in the index
-func (fi *fileIndex) Hashes() []plumbing.Hash {
-	hashes := make([]plumbing.Hash, fi.fanout[0xff])
-	for i := 0; i < fi.fanout[0xff]; i++ {
-		offset := fi.oidLookupOffset + int64(i)*20
-		if n, err := fi.reader.ReadAt(hashes[i][:], offset); err != nil || n < 20 {
-			return nil
-		}
-	}
-	return hashes
-}
+package commitgraph
+
+import (
+	"bytes"
+	"crypto"
+	encbin "encoding/binary"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/hash"
+	"github.com/go-git/go-git/v5/utils/binary"
+)
+
+var (
+	// ErrUnsupportedVersion is returned by OpenFileIndex when the commit graph
+	// file version is not supported.
+	ErrUnsupportedVersion = errors.New("unsupported version")
+	// ErrUnsupportedHash is returned by OpenFileIndex when the commit graph
+	// hash function is not supported. Currently only SHA-1 is defined and
+	// supported
+	ErrUnsupportedHash = errors.New("unsupported hash algorithm")
+	// ErrMalformedCommitGraphFile is returned by OpenFileIndex when the commit
+	// graph file is corrupted.
+	ErrMalformedCommitGraphFile = errors.New("malformed commit graph file")
+
+	commitFileSignature = []byte{'C', 'G', 'P', 'H'}
+	chunkSignatures     = []byte("OIDF OIDL CDAT GDA2 GDO2 EDGE BIDX BDAT BASE \000\000\000\000")
+
+	parentNone        = uint32(0x70000000)
+	parentOctopusUsed = uint32(0x80000000)
+	parentOctopusMask = uint32(0x7fffffff)
+	parentLast        = uint32(0x80000000)
+)
+
+const commitDataSize = 16
+
+// ChunkType represents the type of a chunk in the commit graph file
+type ChunkType int
+
+const (
+	OIDFanoutChunk ChunkType = iota
+	OIDLookupChunk
+	CommitDataChunk
+	GenerationDataChunk
+	GenerationDataOverflowChunk
+	ExtraEdgeListChunk
+	BloomFilterIndexChunk
+	BloomFilterDataChunk
+	BaseGraphsListChunk
+	ZeroChunk
+)
+
+// Signature returns the byte signature for the chunk type
+func (ct ChunkType) Signature() []byte {
+	if ct >= BaseGraphsListChunk || ct < 0 {
+		return chunkSignatures[ZeroChunk*5 : ZeroChunk*5+4]
+	}
+	return chunkSignatures[ct*5 : ct*5+4]
```

Comment:
This `5` and potentially `4` could become constants, so it is easier to grasp the logic.

Reply:
done